### PR TITLE
Implementing slugs for routing for members and characters. 

### LINKED ISSUE #5: Character and member names with forward slashes are breaking pages.
The pages for viewing members and characters escape most of the characters in a name, but aren't escaping slashes. This is causing an invalid URL and a 404.

Short term fix: Make sure slashes are escaped.

Long term fix: Implement slugs for character and member pages.

A slug is when a name is changed to be more friendly to the system. For example "My Friend's Party" becomes "my-friends-party"

### PR DESCRIPTION
**Pull request by Pyrodius0**
Implemented slugs for user and for characters.

This solves the bug where character routing would break.

Added database\hacks\character/memberSlugConversion.php files to track what needs to be executed to convert slugs.

fixes #5 
resolves #5 
closes #5

### CODE REVIEW

File: database/migrations/2020_08_18_032818_add_member_character_slug.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('characters', function (Blueprint $table) {
+            $table->string('slug')->nullable()->after('name');
+        });
+
+        Schema::table('members', function (Blueprint $table) {
+            $table->string('slug')->nullable()->after('username');
```
Comment (Lemmings19) [inline database/migrations/2020_08_18_032818_add_member_character_slug.php:21]: We're going to use this slug to lookup members from the database, so I think we should index both of these new slugs.

The slug function in PHP also limits the slug's length to 50 characters, and that's loads for a username. I think we can also set these fields to be 50 characters for faster indexing+lookup (I am sort of assuming it's faster).

These fields are also required to load a page for either member or character, so we shouldn't allow null on them. (`username` and `name` also have NOT NULL set to true)

```
$table->string('slug', 50)->after('username')->index();
```

Comment (Lemmings19) [reply database/migrations/2020_08_18_032818_add_member_character_slug.php:21]: Guilds and instances have their slugs set to 255 characters, so I think there's an opportunity to optimize those fields too at some point.

**Comment by Pyrodius0** [comment]
> This looks really good, thanks for doing this! I'm waiting for the migration to get sorted out before I run it locally and give it a test. (see comment attached to the migration)
> 
> Other than that, I don't see any issues!

I updated the migration as requested. I updated the same migration  and wiped my database and remigrated to test everything. 

Looks good.

File: database/migrations/2020_08_18_032818_add_member_character_slug.php (hunk 1 above)
Comment (Pyrodius0) [reply database/migrations/2020_08_18_032818_add_member_character_slug.php]: updated as per your feedback.

File: database/migrations/2020_08_18_032818_add_member_character_slug.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+     */
+    public function down()
+    {
+        Schema::table('characters', function (Blueprint $table) {
+            //
+        });
+
+        Schema::table('members', function (Blueprint $table) {
+            //
+        });
```
Comment (Lemmings19) [inline database/migrations/2020_08_18_032818_add_member_character_slug.php:38]: Just a note that these don't do anything unless we program them to undo the changes we made. (but we can skip it for now)

`php artisan migrate:rollback` can then be run to undo a migration.

**Comment by Lemmings19** [comment]
Trying to figure out how to edit my review or open a new one. Ignore any weird stuff regarding reviews.

**Comment by Lemmings19** [comment]
Need to update character and member links in the dashboard: https://github.com/thatsmybis/thatsmybis/blob/master/resources/views/dashboard.blade.php#L26

File: resources/views/member/show.blade.php [hunk 3]
```diff
@@ -55,7 +55,7 @@
                                     @include('partials/item', ['wowheadLink' => false])
                                     <small class="text-muted">
                                         on
-                                        <a href="{{route('character.show', ['guildSlug' => $guild->slug, 'name' => $character->name]) }}"
+                                        <a href="{{route('character.show', ['guildSlug' => $guild->slug, 'nameSlug' => $character->slug]) }}"
```
Comment (Lemmings19) [inline resources/views/member/show.blade.php:58]: Looks like an existing bug, but `$character->slug` needs to be changed to `$itemCharacter->slug`

**Comment by Lemmings19** [comment]
Two changes needed (dashboard, `itemCharacter`), then it should be good for a merge.

File: database/migrations/2020_08_18_032818_add_member_character_slug.php (hunk 2 above)
Comment (Pyrodius0) [reply database/migrations/2020_08_18_032818_add_member_character_slug.php:38]: yeah, I was just stubbing it for later if we need.

**Comment by Pyrodius0** [comment]
> Need to update character and member links in the dashboard: https://github.com/thatsmybis/thatsmybis/blob/master/resources/views/dashboard.blade.php#L26

Not sure what we are missing, but I already updated that. 

image

File: resources/views/member/show.blade.php (hunk 3 above)
Comment (Pyrodius0) [reply resources/views/member/show.blade.php:58]: Updated

File: resources/views/dashboard.blade.php [hunk 4]
```diff
@@ -23,14 +23,14 @@
                     @foreach ($user->members as $member)
                         <li class="bg-lightest mt-3 mb-3 p-3">
                             <h2>
-                                <a href="{{ route('member.show', ['guildSlug' => $member->guild->slug, 'username' => $member->username]) }}" class="text-uncommon font-weight-medium">
+                                <a href="{{ route('member.show', ['guildSlug' => $member->guild->slug, 'usernameSlug' => $member->slug]) }}" class="text-uncommon font-weight-medium">
```
Comment (Pyrodius0) [inline resources/views/dashboard.blade.php:26]: @Lemmings19  I think you were looking at the wrong branch.

File: resources/views/dashboard.blade.php [hunk 5]
```diff
@@ -23,14 +23,14 @@
...
-                                <a href="{{ route('member.show', ['guildSlug' => $member->guild->slug, 'username' => $member->username]) }}" class="text-uncommon font-weight-medium">
+                                <a href="{{ route('member.show', ['guildSlug' => $member->guild->slug, 'usernameSlug' => $member->slug]) }}" class="text-uncommon font-weight-medium">
                                     &lt;{{ $member->guild->name }}&gt;
                                 </a>
                             </h2>
                             <ul class="list-inline">
                                 @foreach ($member->characters as $character)
                                     <li class="list-inline-item bg-tag rounded pt-0 pl-2 pb-1 pr-2 m-2">
-                                        <a href="{{route('character.show', ['guildSlug' => $member->guild->slug, 'name' => $character->name]) }}"
+                                        <a href="{{route('character.show', ['guildSlug' => $member->guild->slug, 'nameSlug' => $character->slug]) }}"
```
Comment (Pyrodius0) [inline resources/views/dashboard.blade.php:33]: @Lemmings19 I think you were looking at the wrong branch.

**Comment by Lemmings19** [comment]
You're right, I probably was on the wrong branch or was missing some changes.

Looks good!

:shipit:
